tests/ga/test_extension.py

```diff
@@ -1464,24 +1464,32 @@ def test_uninstall_failure(self, patch_get_uninstall_command, *args):
         self._assert_no_handler_status(protocol.report_vm_status)
 
     @patch('azurelinuxagent.ga.exthandlers.HandlerManifest.get_update_command')
-    def test_upgrade_failure(self, patch_get_update_command, *args):
+    def test_extension_upgrade_failure_when_update_fails(self, patch_get_update_command, *args):
         """
-        Extension upgrade failure should not be retried
+        When the update command of the new extension fails, it should result in the new extension removed and the
+        old extension disabled. On the next goal state, the entire upgrade scenario should be retried (once),
+        meaning the new extension is downloaded again and update is called again.
```

Comment:
If there's some bug in the update.cmd of the extension, wont this cause the agent to download and delete the zip file at every GS? Aesthetically that would make sense since the code is buggy, but it would still cause extra network bandwidth for the customer. Should we handle this scenario?

Reply:
Initially I was thinking along those lines, too. But since now we go in the loop only on new goal state, my thought was that re-downloading was ok. What do you guys think? @pgombar @vrdmr

Reply:
My concern is mostly about those extensions that the customer doesnt even want or know about (like OMS) and which are extremely buggy. Extensions like this would cause unnecessary traffic which might make them mad. Not sure if that is still part of the scope of this PR

Reply:
> wont this cause the agent to download and delete the zip file at every GS?

The zip files for the new update would still be downloaded but only one-time per GS (not infinite). I feel it is okay. Yes, our downloads might increase in a fast-changing VM, but still would be far less than the earlier problem (similar to what @narrieta said)

The bug of infinite that came back in 2.2.44 was due to the issue that cleanup code does not have the ExtHandler models of the updated extension and thus assumes it is an outdated version. And there is no easy way to fix that problem if we still want to allow infinite attempts.

Reply:
I think it's ok to have the extension get downloaded once per goal state since we got rid of the upgrade loop.